### PR TITLE
Fix: get stuck when load extension in the concurrency environment

### LINKED ISSUE #183: ExtensionDirectory gets stuck when load extension under the concurrency environment
#180

### PR DESCRIPTION
**Pull request by onewe**
- Add a new struct called LoadExtensionPromise
- Remove async modifier in ExtensionDirectory

Close #183

### CODE REVIEW

File: dubbo/src/extension/mod.rs [hunk 1]
```diff
@@ -141,61 +141,95 @@ impl ExtensionDirectory {
...
-            semaphore: Arc::new(Semaphore::new(0)),
+            resolver: Arc::new(RwLock::new(resolver)),
         }
     }
 
-    fn get_extension(&self) -> Option<T> {
-        self.extension.as_ref().as_ref().map(|a| a.clone())
-    }
-
     pub(crate) async fn resolve(&mut self) -> Result<T, StdError> {
```
Comment (kwsc98) [inline dubbo/src/extension/mod.rs:204]: Can be simplified like
```rust
    pub(crate) async fn resolve(&mut self) -> Result<T, StdError> {
        // get read lock
        let resolver_read_lock = self.resolver.read().await;
        // if extension is not None, return it
        if let Some(extension) = resolver_read_lock.resolved_data() {
            return Ok(extension);
        }
        drop(resolver_read_lock);
        let mut resolver_write_lock = self.resolver.write().await;
        Ok(match resolver_write_lock.resolved_data() {
            Some(data) => data,
            None => resolver_write_lock.resolve().await?,
        })
    }
```

**Review by kwsc98 — COMMENTED** [review]
可以简化一下

**Comment by kwsc98** [comment]
good work !

**Comment by yang20150702** [comment]
Good work!

**Pull request merged**